Feat/preflight add `logLevel` option to silence CORS preflight logs

gulbaki opened the pull request:
This PR introduces an optional logLevel field to FastifyCorsOptions, letting users override the Fastify logger level for the internal OPTIONS * pre-flight route.
- Adds logLevel to default options and TypeScript typings
- Applies level when registering the pre-flight route
- New unit test verifies OPTIONS logs are suppressed, GET logs remain
- README/types comments updated accordingly

fixes: #320 




#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/main/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/main/CODE_OF_CONDUCT.md)

gulbaki committed:
feat: add logLevel option for CORS preflight requests

gulbaki committed:
test(cors): add test for logs

gulbaki committed:
feat: update documentation

jean-michelet commented on index.js:
```diff
@@ -72,7 +78,8 @@ function fastifyCors (fastify, opts, next) {
   // This route simply enables fastify to accept preflight requests.
-  fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
+
+  fastify.options('*', { schema: { hide: hideOptionsRoute }, ...(routeLogLevel !== null && { logLevel: routeLogLevel }) }, (req, reply) => {
```
```suggestion
  fastify.options('*', { schema: { hide: hideOptionsRoute }, routeOptions) }, (req, reply) => {
```

Maybe we should prepare a `routeOption` (or more explicit name) object with default to `{}`.

jean-michelet reviewed (commented):
I’m not a user of `logLevel`, but based on the [Fastify docs](https://fastify.dev/docs/latest/Reference/Routes/#custom-log-level) about it, it looks like `logLevel`, when passed in plugin options, applies to the entire plugin. 

In your implementation, however, it only affects the `OPTIONS` route used for preflight requests.

Given that, a more appropriate name for this option might be `preflightLogLevel` or `optionsRouteLogLevel`.

climba03003 commented:
> I’m not a user of logLevel, but based on the [Fastify docs](https://fastify.dev/docs/latest/Reference/Routes/#custom-log-level) about it, it looks like logLevel, when passed in plugin options, applies to the entire plugin.

`logLevel` only works when the plugin is encapsulated. So, in this plugin the `logLevel` is ignored by the core and need to apply manually.

climba03003 commented:
The change below should be enough and there is no need on types change. 
The core types already include the `logLevel` for you.

```diff
diff --git a/index.js b/index.js
index 94c1dd2..7ec92a2 100644
--- a/index.js
+++ b/index.js
@@ -46,6 +46,7 @@ function fastifyCors (fastify, opts, next) {
   fastify.decorateRequest('corsPreflightEnabled', false)

   let hideOptionsRoute = true
+  let logLevel
   if (typeof opts === 'function') {
     handleCorsOptionsDelegator(opts, fastify, { hook: defaultOptions.hook }, next)
   } else if (opts.delegator) {
@@ -53,6 +54,7 @@ function fastifyCors (fastify, opts, next) {
     handleCorsOptionsDelegator(delegator, fastify, options, next)
   } else {
     if (opts.hideOptionsRoute !== undefined) hideOptionsRoute = opts.hideOptionsRoute
+    if (opts.logLevel !== undefined) logLevel = opts.logLevel
     const corsOptions = normalizeCorsOptions(opts)
     validateHook(corsOptions.hook, next)
     if (hookWithPayload.indexOf(corsOptions.hook) !== -1) {
@@ -72,7 +74,7 @@ function fastifyCors (fastify, opts, next) {
   // remove most headers (such as the Authentication header).
   //
   // This route simply enables fastify to accept preflight requests.
-  fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
+  fastify.options('*', { schema: { hide: hideOptionsRoute }, logLevel }, (req, reply) => {
     if (!req.corsPreflightEnabled) {
       // Do not handle preflight requests if the origin option disabled CORS
       reply.callNotFound()
```

gulbaki commented on index.js:
```diff
@@ -46,6 +47,11 @@ function fastifyCors (fastify, opts, next) {
   fastify.decorateRequest('corsPreflightEnabled', false)
 
   let hideOptionsRoute = true
+  let routeLogLevel = null
+
+  if (opts && typeof opts === 'object' && opts.logLevel !== undefined) {
+    routeLogLevel = opts.logLevel
+  }
```
Yes, I checked, and I even noticed it in the following case.
hideOptionsRoute won't work here either if it's used together with a delegator.
` if (opts.hideOptionsRoute !== undefined) hideOptionsRoute = opts.hideOptionsRoute`

```
await app.register(cors, {
    delegator: () => ({ origin: '*' }),
    hideOptionsRoute: true 
  })
```
The delegator will override this setting maybe it's not very important, I'm not sure.

climba03003 commented:
> @climba03003 I think, if I put the if condition there as you said this code will not work ` await app.register(cors, { delegator: () => ({ origin: '*' }), logLevel: 'silent' })`

Same argument apply to `hideOptionsRoute`, just move those outside the else block.
I am also thinking should we allows, or even deprecate the function feature and use option only.
```js
function delegator() {}
delegator.logLevel = 'info'
delegator.hideOptionsRoute = true

fastify.register(CORS, (instance) => delegator)
```

gulbaki committed:
feat(test):logLevel, hideOptionsRoute,  were tested with the delegator in the usage situation.

gulbaki committed:
feat(cors): refactor logLevel handling and streamline options processing

gulbaki commented:
> > @climba03003 I think, if I put the if condition there as you said this code will not work ` await app.register(cors, { delegator: () => ({ origin: '*' }), logLevel: 'silent' })`
> 
> Same argument apply to `hideOptionsRoute`, just move those outside the else block. I am also thinking should we allows, or even deprecate the function feature and use option only.
> 
> ```js
> function delegator() {}
> delegator.logLevel = 'info'
> delegator.hideOptionsRoute = true
> 
> fastify.register(CORS, (instance) => delegator)
> ```
I have refactored the code as you suggested (logLevel, hideOptionsRoute). I also added tests that cover delegator usage for both.
However, regarding the part:
`I am also thinking should we allow, or even deprecate the function feature and use option only.`
If there's anything I should do about this, please let me know . I’d be happy to take care of it.

mcollina reviewed (approved):
lgtm

climba03003 reviewed (changes requested):
Should be the last change before land.

gulbaki committed:
feat(cors): remove logLevel option from default options and update type definition
